client/X11/xf_cliprdr.c

```diff
@@ -1663,6 +1928,334 @@ static UINT xf_cliprdr_clipboard_file_range_failure(wClipboardDelegate* delegate
 	return clipboard->context->ClientFileContentsResponse(clipboard->context, &response);
 }
 
+static void xf_cliprdr_fuse_getattr(fuse_req_t req, fuse_ino_t ino, struct fuse_file_info* fi)
+{
+	double timeout = 0;
+	struct stat stbuf;
+
+	xfCliprdrFuseInode* node;
+	xfClipboard* clipboard = (xfClipboard*)fuse_req_userdata(req);
+
+	ArrayList_Lock(clipboard->ino_list);
+	node = (xfCliprdrFuseInode*)ArrayList_GetItem(clipboard->ino_list, ino - 1);
+	ArrayList_Unlock(clipboard->ino_list);
+	if (!node)
+	{
+		fuse_reply_err(req, ENOENT);
+	}
+	else
+	{
+		memset(&stbuf, 0, sizeof(stbuf));
+		stbuf.st_ino = ino;
+		stbuf.st_mode = node->st_mode;
+		stbuf.st_mtime = node->st_mtim.tv_sec;
+		if (ino == FUSE_ROOT_ID)
+		{
+			stbuf.st_nlink = 2;
+		}
+		else
+		{
+			stbuf.st_nlink = 1;
+		}
+		fuse_reply_attr(req, &stbuf, timeout);
+	}
+}
+
+static void xf_cliprdr_fuse_readdir(fuse_req_t req, fuse_ino_t ino, size_t size, off_t off,
+                                    struct fuse_file_info* fi)
+{
+	size_t count;
+	size_t index;
+	size_t child_ino;
+	size_t direntry_len;
+	char* buf = (char*)calloc(size, sizeof(char));
```

Comment:
It would be nice to check that this allocation succeeds.

Also, this buffer currently leaks if we exit early with ENOENT or ENOTDIR, or return an empty file list. I believe the buffer may be allocated only once we're sure that there is some non-empty data that we want to return.